samcli/lib/list/testable_resources/testable_resources_producer.py

```diff
@@ -0,0 +1,367 @@
+                                deployed_resource["PhysicalResourceId"], stages
+                            )
+                        paths_and_methods = get_methods_and_paths(deployed_resource["LogicalResourceId"], stacks[0])
+
+                    testable_resource_data = TestableResDef(
+                        LogicalResourceId=deployed_resource["LogicalResourceId"],
+                        PhysicalResourceId=deployed_resource["PhysicalResourceId"],
+                        CloudEndpointOrFURL=endpoint_function_url,
+                        Methods=paths_and_methods,
+                    )
+                    testable_resources_list.append(dataclasses.asdict(testable_resource_data))
+                    seen_testable_resources.add(deployed_resource["LogicalResourceId"])
+            for local_resource in stacks[0].resources:
+                local_resource_type = stacks[0].resources[local_resource]["Type"]
+                paths_and_methods = "-"
+                if local_resource_type in testable_resource_types and local_resource not in seen_testable_resources:
+                    if local_resource_type in ("AWS::ApiGateway::RestApi", "AWS::ApiGatewayV2::Api"):
+                        paths_and_methods = get_methods_and_paths(local_resource, stacks[0])
+                    testable_resource_data = TestableResDef(
+                        LogicalResourceId=local_resource,
+                        PhysicalResourceId="-",
+                        CloudEndpointOrFURL="-",
+                        Methods=paths_and_methods,
+                    )
+                    testable_resources_list.append(dataclasses.asdict(testable_resource_data))
+        else:
+            testable_resources_list = get_local_testable_resources(stacks, testable_resource_types)
+        mapped_output = self.mapper.map(testable_resources_list)
+        self.consumer.consume(mapped_output)
+
+
+def validate_stack(stacks: list):
+    """
+    Checks if the stack non-empty and contains stack resources and raises exceptions accordingly
+
+    Parameters
+    ----------
+    stacks: list
+        A list containing the stack
+    """
+    if not stacks or not stacks[0].resources:
+        raise SamListLocalResourcesNotFoundError(msg="No local resources found.")
+
+
+def get_local_testable_resources(stacks: list, testable_resource_types: set) -> list:
+    """
+    Gets a list of local testable resources based on the local stack
+
+    Parameters
+    ----------
+    stacks: list
+        A list containing the stack
+    testable_resource_types: set
+        A set of resources types that should be displayed by testable resources
+
+    Returns
+    -------
+    testable_resources_list: list
+        A list containing the testable resources and their information
+    """
+    testable_resources_list = []
+    paths_and_methods: Any
+    for local_resource in stacks[0].resources:
+        local_resource_type = stacks[0].resources[local_resource]["Type"]
```

Comment:
lets use .get(x, {}) wherever you are accessing the dictionary elements, so that we dont accidentally run into key errors.

Reply:
Changed to using this